livesync_citadel/indico_livesync_citadel/backend.py

```diff
@@ -0,0 +1,111 @@
+# This file is part of the Indico plugins.
+# Copyright (C) 2002 - 2020 CERN
+#
+# The Indico plugins are free software; you can redistribute
+# them and/or modify them under the terms of the MIT License;
+# see the LICENSE file for more details.
+
+from __future__ import unicode_literals
+
+import json
+
+import requests
+from werkzeug.urls import url_join
+
+from indico.modules.attachments import Attachment
+from indico.modules.events import Event
+from indico.modules.events.contributions import Contribution
+from indico.modules.events.contributions.models.subcontributions import SubContribution
+from indico.modules.events.notes.models.notes import EventNote
+
+from indico_livesync import LiveSyncBackendBase, SimpleChange, Uploader
+from indico_livesync_citadel.models.search_id_map import EntryType, LiveSyncCitadelSearchAppIdMap
+from indico_livesync_citadel.schemas import (AttachmentSchema, ContributionSchema, EventNoteSchema, EventSchema,
+                                             SubContributionSchema)
+
+
+class LiveSyncCitadelUploaderError(Exception):
+    pass
+
+
+class LiveSyncCitadelUploader(Uploader):
+    def __init__(self, *args, **kwargs):
+        from indico_livesync_citadel.plugin import LiveSyncCitadelPlugin
+
+        super(LiveSyncCitadelUploader, self).__init__(*args, **kwargs)
+
+        self.search_app = LiveSyncCitadelPlugin.settings.get('search_backend_url')
+        self.tika_server = LiveSyncCitadelPlugin.settings.get('tika_server')
+        self.endpoint_url = url_join(self.search_app, 'api/records')
+        self.headers = {
+            'Content-Type': 'application/json',
+            'Accept': 'application/json',
+            'Authorization': 'Bearer {}'.format(LiveSyncCitadelPlugin.settings.get('search_backend_token'))
+        }
+
+    def get_jsondata(self, obj):
+        # TODO: should "schemas/indico" be another setting?
+        if isinstance(obj, Event):
+            event_data = EventSchema().dump(obj)
+            event_data['$schema'] = url_join(self.search_app, 'schemas/indico/events_v1.0.0.json')
```

Comment:
I have been thinking whether this `schemas/indico` part should be a plugin-specific setting.

Reply:
What we talked about not too long ago: When would you ever need other schemas, without being the developer/maintainer of this plugin and making a new release of it as well?

Reply:
I'd also remove the version number from the file name. Replacing a file w/o renaming is going to give you a much nicer git history!

Reply:
> What we talked about not too long ago: When would you ever need other schemas, without being the developer/maintainer of this plugin and making a new release of it as well?

I don't quite understand what you meant in this comment.

Reply:
I don't understand when someone who is just managing an Indico instance (and running an unmodified citadel instance) would want to change the schemas being used.

Reply:
I still don't understand what you mean. This `$schema` property is used by ES to find out which index to use when creating new objects.

Reply:
Oh, I always thought `schemas/indico/events_v1.0.0.json` was a file included in the plugin with some kind of metadata...
So that's not the case? Does the `.json` in there make sense then? If it's just an ES schema name, then it being configurable seems much more reasonable. And maybe we can have a clearer default? Maybe something like `indico-events`  or `indico-events-v1` if versioning is needed.

However, does changing the schema name require  full resync? In that case I wonder if you'd ever want to change the schema name once there's data...

Reply:
> Oh, I always thought schemas/indico/events_v1.0.0.json was a file included in the plugin with some kind of metadata...

Those files only exists on the ES instance.

> So that's not the case? Does the .json in there make sense then?

Probably, since this is jsonschema.

> Maybe something like indico-events or indico-events-v1 if versioning is needed.

Will check with the CERN Search developer. As far as I can tell there are some possibilities of changing that, but I need to clarify this.

> However, does changing the schema name require full resync? 

Adding new fields requires a full resync. Changing name probably not, but it is also something I need to confirm.

Reply:
> Maybe something like indico-events or indico-events-v1 if versioning is needed

Apparently the versioning part is needed in the schema name.

So now, I am wondering whether there should be a setting for the schema names. I think that in case of Indico they will never change anyway. And if they do we can easily modify the code.

Reply:
I'd keep them in the code then. Otherwise it confuses indico administrators who are not that familiar with stuff beyond what's in the setup guide. And if there is ever a schema change we probably want to bump the plugin version and be very clear about what they need to do.